File: evals/refs/fft.zig
const std = @import("std");

const C = struct { re: f64, im: f64 };

fn cadd(a: C, b: C) C { return .{ .re = a.re + b.re, .im = a.im + b.im }; }
fn csub(a: C, b: C) C { return .{ .re = a.re - b.re, .im = a.im - b.im }; }
fn cmul(a: C, b: C) C {
    return .{ .re = a.re * b.re - a.im * b.im, .im = a.re * b.im + a.im * b.re };
}

fn fft(x: []C, allocator: std.mem.Allocator) !void {
    const n = x.len;
    if (n == 1) return;
    const half = n / 2;
    const e = try allocator.alloc(C, half);
    const o = try allocator.alloc(C, half);
    var k: usize = 0;
    while (k < half) : (k += 1) {
        e[k] = x[2 * k];
        o[k] = x[2 * k + 1];
    }
    try fft(e, allocator);
    try fft(o, allocator);
    k = 0;
    while (k < half) : (k += 1) {
        const theta = -2.0 * std.math.pi * @as(f64, @floatFromInt(k)) / @as(f64, @floatFromInt(n));
        const w: C = .{ .re = @cos(theta), .im = @sin(theta) };
        const t = cmul(w, o[k]);
        x[k] = cadd(e[k], t);
        x[k + half] = csub(e[k], t);
    }
}

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();
    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;
    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);
    var out_buf: [16384]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;
    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var c: usize = 0;
    while (c < t) : (c += 1) {
        const n = try std.fmt.parseInt(usize, it.next().?, 10);
        const x = try arena.alloc(C, n);
        var i: usize = 0;
        while (i < n) : (i += 1) {
            const v = try std.fmt.parseFloat(f64, it.next().?);
            x[i] = .{ .re = v, .im = 0.0 };
        }
        try fft(x, arena);
        i = 0;
        while (i < n) : (i += 1) {
            try w.print("{d:.4} {d:.4}\n", .{ x[i].re, x[i].im });
        }
    }
    try w.flush();
}

File: evals/refs/fft.rs
use std::io::{self, Read, Write, BufWriter};

#[derive(Copy, Clone)]
struct C { re: f64, im: f64 }
impl std::ops::Add for C { type Output = C; fn add(self, r: C) -> C { C { re: self.re + r.re, im: self.im + r.im } } }
impl std::ops::Sub for C { type Output = C; fn sub(self, r: C) -> C { C { re: self.re - r.re, im: self.im - r.im } } }
impl std::ops::Mul for C { type Output = C; fn mul(self, r: C) -> C { C { re: self.re*r.re - self.im*r.im, im: self.re*r.im + self.im*r.re } } }

fn fft(x: &mut [C]) {
    let n = x.len();
    if n == 1 { return; }
    let mut e: Vec<C> = (0..n/2).map(|k| x[2*k]).collect();
    let mut o: Vec<C> = (0..n/2).map(|k| x[2*k+1]).collect();
    fft(&mut e); fft(&mut o);
    for k in 0..n/2 {
        let theta = -2.0 * std::f64::consts::PI * (k as f64) / (n as f64);
        let w = C { re: theta.cos(), im: theta.sin() };
        let t = w * o[k];
        x[k] = e[k] + t;
        x[k + n/2] = e[k] - t;
    }
}

fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut it = s.split_ascii_whitespace();
    let t: usize = it.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let n: usize = it.next().unwrap().parse().unwrap();
        let mut x: Vec<C> = (0..n).map(|_| {
            let v: f64 = it.next().unwrap().parse().unwrap();
            C { re: v, im: 0.0 }
        }).collect();
        fft(&mut x);
        for c in x {
            writeln!(out, "{:.4} {:.4}", c.re, c.im).unwrap();
        }
    }
}
